bluemira/equilibria/fem_fixed_boundary/fem_magnetostatic_2D.py

```diff
@@ -183,3 +200,255 @@ def calculate_b(self):
+            still 0 and pprime and ffprime are, then, not defined)
+
+        Returns
+        -------
+        g: callable
+            Source current to solve the magnetostatic problem
+        """
+        dx = dolfin.Measure("dx", domain=self.mesh)
+        area = dolfin.assemble(dolfin.Constant(1) * dx())
+        j_target = curr_target / area
+
+        def g(x):
+            if self.psi_ax == 0:
+                return j_target
+            else:
+                r = x[0]
+                x_psi = self.psi_norm_2d(x)
+
+                if callable(pprime):
+                    a = r * pprime(x_psi)
+                else:
+                    a = r * pprime
+
+                if callable(ffprime):
+                    b = 1 / MU_0 / r * ffprime(x_psi)
+                else:
+                    b = 1 / MU_0 / r * ffprime
+
+                return self.k * (a + b)
+
+        return g
+
+    def _create_g(
+        self,
+        pprime: Union[Callable, float],
+        ffprime: Union[Callable, float],
+        curr_target: float,
+    ):
+        """
+        Return the density current DOLFIN function given pprime and ffprime.
+
+        Parameters
+        ----------
+        pprime: Union[callable, float]
+            pprime as function of psi_norm (1-D function)
+        ffprime: Union[callable, float]
+            ffprime as function of psi_norm (1-D function)
+        curr_target: float
+            Target current (also used to initialize the solution in case self.psi is
+            still 0 and pprime and ffprime are, then, not defined)
+
+        Returns
+        -------
+        g: callable
+            Source current to solve the magnetostatic problem
+        """
+        myfunc = self._create_g_func(pprime, ffprime, curr_target)
+        func = ScalarSubFunc(myfunc)
+        return func
+
+    def _calculate_curr_tot(self):
+        """Calculate the total current into the domain"""
+        dx = dolfin.Measure("dx", domain=self.mesh)
+        return dolfin.assemble(self.g * dx())
+
+    def _update_curr(self, curr_target):
+        self.k = 1
+        self.k = curr_target / self._calculate_curr_tot()
+
+    def _plot_current_iteration(self, points, i_iter):
+        curr_data = np.array([self.g_func(p) for p in points])
+        self._plot_array(
+            points, curr_data, f"J current at iteration {i_iter}", contour=False
+        )
+
+    def _plot_array(self, points, array, title, contour=True):
+        plot_scalar_field(
+            points[:, 0],
+            points[:, 1],
+            array,
+            levels=20,
+            axis=None,
+            tofill=True,
+            contour=contour,
+        )
+        plt.title(title)
+        plt.show()
+
+    def solve(
+        self,
+        pprime,
+        ffprime,
+        curr_target,
+        dirichlet_bc_function=None,
+        dirichlet_marker=None,
+        neumann_bc_function=None,
```

Comment:
I don't see anybody changing these defaults any time soon..! Obviously I've only been running the one case and never changing this, so it's as you see fit.